modules/core/src/main/scala/iota/internal/toolbelts.scala

```diff
@@ -0,0 +1,166 @@
+
+  lazy val showTrees =
+    !c.inferImplicitValue(typeOf[debug.optionTypes.ShowTrees], true).isEmpty
+
+  def foldAbort[F[_]: Foldable, T](
+    either: Either[F[String], Tree],
+    isImplicit: Boolean = false
+  ): c.Expr[T] =
+    either fold (
+      errors => {
+        val error = errors.toList.mkString(", and\n")
+        if (isImplicit && showAborts) c.echo(c.enclosingPosition, error)
+        c.abort(c.enclosingPosition, error)
+      },
+      tree  => c.Expr[T](tree))
+
+  def memoize[A, B](cache: IotaMacroToolbelt.Cache)(
+    a: A, f: A => B
+  ): B = cache.underlying.synchronized {
+    cache.underlying.get(a) match {
+      case Some(b: B @unchecked) =>
+        if (showCache)
+          c.echo(c.enclosingPosition, s"ShowCache: $b cached result $b")
+        b
+      case _ =>
+        val b = f(a)
+        cache.underlying += ((a, b))
+        if (showCache)
+          c.echo(c.enclosingPosition, s"ShowCache: $a computed result $b")
+        b
+    }
+  }
+
+  def memoizedTListTypes(tpe: Type): Either[Id[String], List[Type]] =
+    memoize(IotaMacroToolbelt.typeListCache)(tpe, tlistTypes)
+
+  def memoizedKListTypes(tpe: Type): Either[Id[String], List[Type]] =
+    memoize(IotaMacroToolbelt.typeListCache)(tpe, klistTypes)
+
+}
+
+object IotaReflectiveToolbelt {
+  def apply[U <: Universe](u: U): IotaReflectiveToolbelt[u.type] =
+    new IotaReflectiveToolbelt[u.type](u)
+
+  def apply(): IotaReflectiveToolbelt[runtimeUniverse.type] =
+    apply(runtimeUniverse)
+}
+
+class IotaReflectiveToolbelt[U <: Universe](override val u: U)
+    extends IotaCommonToolbelt { override type Uu = U }
+
+private[internal] sealed abstract class IotaCommonToolbelt {
+  type Uu <: Universe
+  val u: Uu
+
+  import u._
+
+  final lazy val TNilTpe  = typeOf[iota.TNil]
+  final lazy val TConsTpe = typeOf[iota.TCons[Nothing, Nothing]].etaExpand.resultType
+
+  final lazy val KNilTpe  = typeOf[iota.KNil]
+  final lazy val KConsTpe = typeOf[iota.KCons[Nothing, Nothing]].etaExpand.resultType
+
+  final lazy val CopTpe   = typeOf[iota.Cop[Nothing]].etaExpand.resultType
+  final lazy val CopKTpe  = typeOf[iota.CopK[Nothing, Nothing]].etaExpand.resultType
+
+  @tailrec
+  final def typeListFoldLeft[A](
+    nilTpe: Type, consTpe: Type
+  )(tpe: Type)(a0: A)(f: (A, Type) => A): Either[String, A] = tpe.dealias match {
+    case TypeRef(_, sym, Nil) if sym.asType.toType <:< nilTpe => Right(a0)
+    case TypeRef(_, sym, head :: tail :: Nil) if sym.asType.toType <:< consTpe =>
+      typeListFoldLeft(nilTpe, consTpe)(tail)(f(a0, head))(f)
+    case _ =>
+      Left(s"Unexpected type ${showRaw(tpe)} when inspecting type list")
+  }
+
+  final def typeListTypes(
+    nilTpe: Type, consTpe: Type
+  )(tpe: Type): Either[Id[String], List[Type]] =
+    typeListFoldLeft(nilTpe, consTpe)(tpe)(List.empty[Type])((acc, t) => t :: acc).map(_.reverse)
+
+  final def tlistTypes(tpe: Type): Either[Id[String], List[Type]] =
```

Comment:
I'd really expect this to use Iota coproducts, but ok, I guess it will do.... 😃 

![image](https://cloud.githubusercontent.com/assets/1645142/25843528/81f8163a-349f-11e7-958e-8c31e3f4776b.png)